apps/core/src/utils/transaction/createTokenTransferTransaction.ts

```diff
@@ -22,43 +21,56 @@ export function createTokenTransferTransaction({
     coins,
     coinType,
     coinDecimals,
-    isPayAllIota,
 }: Options) {
     const tx = new Transaction();
 
-    if (isPayAllIota && coinType === IOTA_TYPE_ARG) {
-        tx.transferObjects([tx.gas], to);
-        tx.setGasPayment(
-            coins
-                .filter((coin) => coin.coinType === coinType)
-                .map((coin) => ({
-                    objectId: coin.coinObjectId,
-                    digest: coin.digest,
-                    version: coin.version,
-                })),
-        );
-
-        return tx;
-    }
+    const totalBalance = coins.reduce((acc, { balance }) => {
+        return BigInt(acc) + BigInt(balance);
+    }, BigInt(0));
 
     const bigIntAmount = parseAmount(amount, coinDecimals);
-    const [primaryCoin, ...mergeCoins] = coins.filter((coin) => coin.coinType === coinType);
+    const isTransferAllObjects = totalBalance === bigIntAmount;
 
     if (coinType === IOTA_TYPE_ARG) {
-        const coin = tx.splitCoins(tx.gas, [bigIntAmount]);
-        tx.transferObjects([coin], to);
+        sendCoins(tx, to, tx.gas, isTransferAllObjects, bigIntAmount);
     } else {
-        const primaryCoinInput = tx.object(primaryCoin.coinObjectId);
-        if (mergeCoins.length) {
-            // TODO: This could just merge a subset of coins that meet the balance requirements instead of all of them.
-            tx.mergeCoins(
-                primaryCoinInput,
-                mergeCoins.map((coin) => tx.object(coin.coinObjectId)),
-            );
-        }
-        const coin = tx.splitCoins(primaryCoinInput, [bigIntAmount]);
-        tx.transferObjects([coin], to);
+        handleCoinTransfer(tx, to, coins, coinType, isTransferAllObjects, bigIntAmount);
     }
 
     return tx;
 }
+
+function handleCoinTransfer(
+    tx: Transaction,
+    to: string,
+    coins: CoinStruct[],
+    coinType: string,
+    isTransferAllObjects: boolean,
+    bigIntAmount: bigint,
+) {
+    const [primaryCoin, ...mergeCoins] = coins.filter((coin) => coin.coinType === coinType);
+    const primaryCoinInput = tx.object(primaryCoin.coinObjectId);
+
+    if (mergeCoins.length) {
+        tx.mergeCoins(
```

Comment:
the TODO comment has been deleted, but it has not been taken into account, right? 
Here all the coins are merged instead of the necessary ones. Is this the behavior we want?

`// TODO: This could just merge a subset of coins that meet the balance requirements instead of all of them.`